Mathlib/RingTheory/Ideal/Basic.lean

```diff
@@ -519,6 +519,26 @@ theorem span_singleton_eq_span_singleton {α : Type u} [CommRing α] [IsDomain 
   apply and_congr <;> rw [span_singleton_le_span_singleton]
 #align ideal.span_singleton_eq_span_singleton Ideal.span_singleton_eq_span_singleton
 
+/-- The equivalence between `R/Rˣ` and the principal ideals of `R` defined by sending the
+class of `x` to the principal ideal generated by `x`. -/
+noncomputable def equivIsPrincipal (R : Type*) [CommRing R] [IsDomain R] :
+    Quotient (MulAction.orbitRel Rˣ R) ≃ {I : Ideal R | Submodule.IsPrincipal I} := by
```

Comment:
Isn't the RHS `Associates R`?

Reply:
And the RHS should maybe be `Subtype` instead.

Reply:
Yes, indeed! It has a monoid structure so I can generalize to `equivIsPrincipal` being a MulEquiv. Thanks!

Reply:
> And the RHS should maybe be `Subtype` instead.

I guess it is better to have function with values in types indeed.

Reply:
RHS does not have a multiplicative structure though. Maybe make it into a `Submonoid (Ideal R)`? And it is even an ordered iso. But I couldn't find a ordered-monoid-iso structure in mathlib.

Reply:
Oh right, it's the ideals that have a structure. Well, I guess that will be for another PR...

Reply:
Maybe make it an order iso? And the inverse is given by `Submodule.IsPrincipal.generator`.

Reply:
Yeah, I was thinking about something like that. But maybe in that case, it would be better to move all the `Submodule.IsPrincipal.generator` stuff in  so that we do not need to import `Mathlib.RingTheory.PrincipalIdealDomain` to have access to the isomorphism but the file `Ideal.Basic` is quite long already... I guess one way to do things would be create a `IsPrincipal` file...

Reply:
Yeah but maybe for another PR.